jobs/build-charms/charms.py

```diff
@@ -39,46 +42,306 @@
+        out = sh.charm.push(dst_path, entity, _out=self._echo)
+        self._echo(f"Pushing   :: {entity:^35} :: returns {out.stdout or out.stderr}")
+
+        # Output includes lots of ansi escape sequences from the docker push,
+        # and we only care about the first line, which contains the url as yaml.
+        out = yaml.safe_load(out.stdout.decode().strip().splitlines()[0])
+        return out["url"]
+
+    def set(self, entity, commit):
+        self._echo(f"Setting {entity} metadata: {commit}")
+        sh.charm.set(entity, f"commit={commit}", _out=self._echo)
+
+    def upload_resource(self, entity, resource_name, resource):
+        # If the resource is a file, populate the path where it was built.
+        # If it's a custom image, it will be in Docker and this will be a no-op.
+        retry_call(
+            sh.charm.attach,
+            fargs=[
+                entity,
+                f"{resource_name}={resource}",
+            ],
+            fkwargs={"_out": self._echo},
+            delay=2,
+            backoff=2,
+            tries=15,
+            exceptions=sh.ErrorReturnCode,
+        )
+
+
+class _CharmHub:
+    STATUS_RESOURCE = re.compile(r"(\S+) \(r(\d+)\)")
+
+    def __init__(self, build_entity):
+        self._echo = build_entity.echo
+
+    @staticmethod
+    def _table_to_list(header, body):
+        if not body:
+            return None
+        rows = []
+        titles = [title for title in re.split(r"\s{2,}", header) if title]
+        for line in body:
+            row, head = {}, line
+            for key in reversed(titles):
+                head, *value = head.strip().rsplit("  ", 1)
+                head, value = (head, value[0]) if value else ("", head.strip())
+                row[key] = value or rows[-1].get(key)
+            rows.append(row)
+        return rows
+
+    @staticmethod
+    def refresh(name, channel=None, architecture=None, base_channel=None):
+        channel = channel or "stable"
+        architecture = architecture or "amd64"
+        base_channel = base_channel or "20.04"
+        data = {
+            "context": [],
+            "actions": [
+                {
+                    "name": name,
+                    "base": {
+                        "name": "ubuntu",
+                        "architecture": architecture,
+                        "channel": base_channel,
+                    },
+                    "channel": channel,
+                    "action": "install",
+                    "instance-key": "charmed-kubernetes/build-charms",
+                }
+            ],
+        }
+        resp = requests.post("https://api.charmhub.io/v2/charms/refresh", json=data)
+        return resp.json()
+
+    def status(self, charm_entity):
+        """Read CLI Table output from charmcraft status and parse."""
+        charm_status = sh.charmcraft.status(charm_entity, _out=self._echo)
+        header, *body = charm_status.stdout.decode().splitlines()
```

Comment:
I am trying some of the parsing locally and the list that is the result of splitlines is empty if _out=self._echo is included in the status sh call. If the call is simply `sh.charmcraft.status(charm_entity) then the list is returned as expected.

Reply:
what if you remove `_out` and replace with `_tee=True`?

```suggestion
    def status(self, charm_entity):
        """Read CLI Table output from charmcraft status and parse."""
        charm_status = sh.charmcraft.status(charm_entity, _tee=True)
        header, *body = charm_status.stdout.decode().splitlines()
```

Reply:
> what if you remove `_out` and replace with `_tee=True`?

The list looks correct with that

Reply:
but again, this is using charmcraft 1.4.0, which might be the issue

Reply:
what if you have both `_tee` and `_out`?

Reply:
now we're talkin. List looks good and the output is shown.